### PR TITLE
[REEF-1420] Dispose IGroupCommClient/Network Service from IMRU tasks

### PR DESCRIPTION
**Pull request by jwang98052**
JIRA: [REEF-1420](https://issues.apache.org/jira/browse/REEF-1420)
This closes

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs [hunk 1]
```diff
@@ -72,6 +72,11 @@ internal sealed class UpdateTaskHost<TMapInput, TMapOutput, TResult> : ITask, IO
         private readonly ManualResetEventSlim _waitToCloseEvent = new ManualResetEventSlim(false);
 
         /// <summary>
+        /// Group Communication client for the task
+        /// </summary>
+        private IGroupCommClient _groupCommunicationsClient;
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs]: `readonly`.

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs [hunk 2]
```diff
@@ -70,6 +70,11 @@ internal sealed class MapTaskHost<TMapInput, TMapOutput> : ITask, IObserver<IClo
         private readonly ManualResetEventSlim _waitToCloseEvent = new ManualResetEventSlim(false);
 
         /// <summary>
+        /// Group Communication client for the task
+        /// </summary>
+        private IGroupCommClient _groupCommunicationsClient;
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs]: `readonly`.

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs [hunk 3]
```diff
@@ -160,6 +166,7 @@ public void OnNext(ICloseEvent closeEvent)
         /// </summary>
         public void Dispose()
         {
+            _groupCommunicationsClient.Dispose();
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs]: Make re-entrant and thread-safe.

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs [hunk 4]
```diff
@@ -180,6 +186,7 @@ public void OnNext(ICloseEvent closeEvent)
         /// </summary>
         public void Dispose()
         {
+            _groupCommunicationsClient.Dispose();
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs]: Make re-entrant and thread-safe.

**Comment by afchung** [comment]
Done with a pass.

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs (hunk 1 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs]: done

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs (hunk 2 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs]: done

**Comment by jwang98052** [comment]
@afchung I have address the review comments. Thanks!

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs [hunk 5]
```diff
@@ -180,6 +191,11 @@ public void OnNext(ICloseEvent closeEvent)
         /// </summary>
         public void Dispose()
         {
+            if (Interlocked.Read(ref _disposed) == 0)
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs]: This does not work, you need to use the result of `Interlocked.Exchange` instead.

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs [hunk 6]
```diff
@@ -160,6 +171,11 @@ public void OnNext(ICloseEvent closeEvent)
         /// </summary>
         public void Dispose()
         {
+            if (Interlocked.Read(ref _disposed) == 0)
```
Comment (afchung) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs]: This does not work, you need to use the result of Interlocked.Exchange instead.

**Comment by afchung** [comment]
@jwang98052 done with another pass.

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs (hunk 6 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/MapTaskHost.cs]: done

File: lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs (hunk 5 above)
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/UpdateTaskHost.cs]: done

**Comment by jwang98052** [comment]
@afchung Thanks for the review. change pushed.

**Comment by afchung** [comment]
I'll merge.

**Pull request closed without merging**